exchange/bidder.go

```diff
@@ -85,7 +86,7 @@ type bidderAdapter struct {
 	Client *http.Client
 }
 
-func (bidder *bidderAdapter) requestBid(ctx context.Context, request *openrtb.BidRequest, name openrtb_ext.BidderName, bidAdjustment float64) (*pbsOrtbSeatBid, []error) {
+func (bidder *bidderAdapter) requestBid(ctx context.Context, request *openrtb.BidRequest, name openrtb_ext.BidderName, bidAdjustment float64, rateConverter *currencies.RateConverter) (*pbsOrtbSeatBid, []error) {
```

Comment:
Yes, that was the purpose of breaking out the `Rates` object. `requestBids()` would user the `rateConverter` once to get the `Rates` object, then use that set of rates for all the conversions. This also benefits performance as you avoid the atomic `load` of the rate data happening more than once in an auction.

Reply:
Sure !
So I introduced a `ConstantRates` which returns 1 if both from and to currencies are the same, otherwise it returns not found.
It allowed to reduce the complexity in the bidder code which receive directly a `Conversions` interface which is either a `ConstantRates` or `Rates`.
The `Rates()` is called once per auction in `exchange.HoldAuction()` to avoid any discrepancies between bidders and the unnecessary multiple fetch per auction.

I kept some tests in `bidder_test` tackling common use cases (checked in `currencies` package and in PBS workflow in bidder just to make sure.

So the bidder code change is minimal but is backed with unit tests and most of the logic is handled in the `Conversions` implementations
Let me know if you guys prefer this implementation (I will be happy to change it otherwise)